Add a metatable for vectors

Desour opened the pull request:
- Goal of the PR:
Make it much easier to use vectors, but still keep backwards compatibility.
Also change the internal representation from `{x=num, ...}` to `{num, ...}`.
- How does the PR work?
There's a metatable that `vector.new` (and any other vector helper) sets for each vector. See also http://www.lua.org/manual/5.1/manual.html#2.8.
- I've added unittests and documentation.
- Closes #7196, closes #9146, there might be more.
- If not a bug fix, why is this PR needed? What usecases does it solve?
This has already been discussed in the issues, but if you didn't follow them, just look at the table below.

old code | new code
----------- | ------
`vector.multiply(vector.add(vector.multiply(a, -1), b), 2)` | `(-a + b) * 2`
`vector.length(a)` | `a:length()`
`local xyz = {"x", "y", "z"} for i = 1, 3 do if v[xyz[i]] == 0 then print("bla") end end` | `for i = 1, 3 do if v[i] == 0 then print("bla") end end`
`if vector.equal(a, b) then` | `if a == b then`

-------------------------

<details>
<summary>Benchmarks</summary>
branches:

- `master`
- `vector_metanonum` (#10330: no array-style vectors)
- `vector_metatable` (this PR)

lua versions:

- luajit: LuaJIT 2.0.5 -- Copyright (C) 2005-2017 Mike Pall. http://luajit.org/
- puclua: Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio

test-cases:

- rayIter:  
  iterate through `vector.rayIter` from `vector_extras` n1 times  
  luajit: n1=100000  
  puclua: n1=50000
- plane:  
  use `vector.plane` from `vector_extras` with the corner positions `{vector.new(n2, 0, 0), vector.new(0, n2, 0), vector.new(0, 0, n2)}`  
  luajit: n2=500  
  puclua: n2=100
- rayIter(new):  
  same as rayIter, but vector helpers in `vector.rayIter` were replaced with new versions (ie. operators and :-syntax)  
  see also `meins_newv` branch of my `vector_extras` fork: https://github.com/Desour/vector_extras  
  (obviously not used with minetest master branch)
- plane(new):  
  the same for plane as rayIter(new) is for rayIter

other remarks:

- multiple numbers in one results field are multiple meassurements
- master does not include any commits that are not in the other branches (both, minetest and vector_extras)

### results:

luajit:

test-case   |master                     |vector_metanonum           |vector_metatable
------------|---------------------------|---------------------------|---------------------------
rayIter     |344.248 ms 338.956 ms      |382.32 ms  353.601 ms      |397.52 ms  395.998 ms
plane       |443.661 ms 443.371 ms      |444.651 ms 444.69 ms       |562.25 ms  552.524 ms
rayIter(new)|-                          |473.845 ms 481.517 ms      |481.585 ms 490.468 ms
plane(new)  |-                          |452.801 ms 457.097 ms      |483.506 ms 489.471 ms

puclua:

test-case   |master                     |vector_metanonum           |vector_metatable
------------|---------------------------|---------------------------|---------------------------
rayIter     |1097.359 ms 1109.879 ms    |1216.742 ms 1223.427 ms    |2284.12 ms  2311.729 ms
plane       |1684.171 ms 1714.625 ms    |1801.363 ms 1838.011 ms    |3845.351 ms 3850.301 ms
rayIter(new)|-                          |1317.91 ms  1331.468 ms    |1927.844 ms 1969.003 ms
plane(new)  |-                          |3057.31 ms  3092.618 ms    |4008.22 ms  4070.576 ms

<details><summary>History</summary>

- 2020-10-05 (23:28): Added benchmarks to top comment.

</details>

<details><summary>Code</summary>

`vector_extras`: https://github.com/HybridDog/vector_extras
fork of `vector_extras` with `meins_newv` branch: https://github.com/Desour/vector_extras

`vector.rayIter`:
https://github.com/HybridDog/vector_extras/blob/8c7aaf6c0bd5f4fa2ae4d41c22454c3f7d4fe42a/init.lua#L101-L131
`vector.plane`:
https://github.com/HybridDog/vector_extras/blob/8c7aaf6c0bd5f4fa2ae4d41c22454c3f7d4fe42a/init.lua#L274-L327

benchmark code:
```lua
local function meassure(name, func, ...)
	print(name..":")
	local time_start = minetest.get_us_time()
	func(...)
	local time_end = minetest.get_us_time()
	print("time taken: " .. ((time_end - time_start) / 1000) .. " ms")
end

local is_luajit = rawget(_G, "jit") ~= nil

print("benchmarking with vector_extras...")
print("Lua: " .. (is_luajit and "LuaJIT" or "PUC Lua"))

local n1 = is_luajit and 100000 or 50000
meassure("rayIter", function()
	local i = 0
	for _ in vector.rayIter(vector.new(1, 2, 3), vector.new(3, 4, 7)) do
		i = i + 1
		if i == n1 then
			break
		end
	end
end)

local n2 = is_luajit and 500 or 100
meassure("plane", vector.plane, {vector.new(n2, 0, 0), vector.new(0, n2, 0), vector.new(0, 0, n2)})

print("benchmarking with vector_extras done.")
```

</details>

</details>

----------------------------

## To do

<!-- This PR is a Ready for Review. -->
This PR is a Work in Progress.

## How to test

- Run the unittests as described in #9184.
- Start a world in devtest game to run the test for json.
- Play with your mods to look for regressions.
- Use the new features in a mod.

sfan5 commented on builtin/common/tests/vector_spec.lua:
```diff
@@ -4,14 +4,14 @@ dofile("builtin/common/vector.lua")
-			assert.same({ x = 3, y = 2, z = 1 }, vector.new({ x = 3, y = 2, z = 1 }))
+			assert.same({ 0, 0, 0 }, vector.new())
+			assert.same({ 1, 2, 3 }, vector.new(1, 2, 3))
+			assert.same({ 3, 2, 1 }, vector.new({ x = 3, y = 2, z = 1 }))
```
should have a test for the metatable

Desour replied:
I thought that, too. But in http://www.lua.org/manual/5.1/manual.html#2.8 section "eq", it says that it already compares metatables in `getcomphandler` each time `equal` is used.

sfan5 replied:
No I meant like: `assert.is_true(getmetatable(vector.new()) ~= nil)`

Desour replied:
Done using the new vector.is_vector.

sfan5 commented on builtin/common/tests/vector_spec.lua:
```diff
@@ -25,6 +25,31 @@ describe("vector", function()
+		assert.equal(13, some_vector.z)
+
+		some_vector[1] = 100
+		assert.equal(100, some_vector[1])
```
maybe test the opposite here: `some_vector.x`

Desour replied:
Done.

sfan5 commented on builtin/common/tests/vector_spec.lua:
```diff
@@ -38,9 +63,45 @@ describe("vector", function()
 
 	it("add()", function()
-		assert.same({ x = 2, y = 4, z = 6 }, vector.add(vector.new(1, 2, 3), { x = 1, y = 2, z = 3 }))
+		assert.equal(vector.new(2, 4, 6), vector.add(vector.new(1, 2, 3), { x = 1, y = 2, z = 3 }))
```
why does `same` not work here?

Desour replied:
Same seems to compare the raw tables.
(I don't think any mod does this / needs to do this.)

Desour replied:
Now that next and pairs is overridden, same would work here.

Desour replied:
But `equal` is fine. (`same` was *needed* in the past.)

sfan5 commented on builtin/common/tests/vector_spec.lua:
```diff
@@ -38,9 +63,45 @@ describe("vector", function()
+		assert.equal(23, vector.length(vector.new(0, 0, -23)))
+	end)
+
+	it("operators", function()
```
Missing a test for `__unm`?

Desour replied:
No, see line 93.

sfan5 commented on builtin/common/tests/vector_spec.lua:
```diff
@@ -38,9 +63,45 @@ describe("vector", function()
+		assert.equal(a + b, b + a)
+		assert.equal(a + 1, 1 + a)
+
+		assert.equal(vector.subtract(a, b), a - b)
```
what about `a:subtract(...)`?

Desour replied:
The tests aren't meant to be covering as they weren't before either.
Testing one function with `:` should suffice.

sfan5 replied:
But it would be nice improve coverage.

Desour replied:
Done.
But I won't add tests for multiply and co., it would blow up this PR even more.

sfan5 commented on builtin/common/vector.lua:
```diff
@@ -1,58 +1,96 @@
+	elseif key == "z" then
+		v[3] = value
+	else
+		error("Invalid key for vector", 2)
```
Not being able to set foreign keys in a vector is a possible backwards compatibility break.

Desour replied:
Done.

sfan5 commented on builtin/common/vector.lua:
```diff
@@ -1,58 +1,96 @@
+		math.floor(v.x + 0.5),
+		math.floor(v.y + 0.5),
+		math.floor(v.z + 0.5)
+	)
```
I worry a little about `vector.new` (... if-else and asserts) being called for every calculation

Desour replied:
Done.

sfan5 commented on builtin/common/vector.lua:
```diff
@@ -82,62 +116,110 @@ function vector.dot(a, b)
+end
+
+function metatable.__unm(v)
+	return v * -1
```
should perhaps be inlined for speed

Desour replied:
Done.

sfan5 commented on doc/lua_api.txt:
```diff
@@ -1379,6 +1379,9 @@ Position/vector
 
     {x=num, y=num, z=num}
 
+    Note: it is recommended to construct a vector rather like this:
```
```suggestion
    Note: it is recommended to construct a vector using the helper function:
```

Desour replied:
Done.

srifqi commented on doc/lua_api.txt:
```diff
@@ -1,4 +1,4 @@
-Minetest Lua Modding API Reference
+rMinetest Lua Modding API Reference
```
unintended change?

Desour replied:
Yes.

Desour commented:
Also added a `vector.is_vector` function, tell me if I should split it into two (or just remove it entirely, the metatable can be gotten from and other vector with `getmetatable`).

Desour commented:
I could increase the speed a little bit by using different functions in `vector.*` and in the metatable.
Tell me if I should do so (and if yes, also tell me whether it should be possible to add an old vector to a nice one). (Such optimizations would still be possible afterwards.)

Desour commented:
Btw. related irc discussion: http://irc.minetest.net/minetest-dev/2020-04-07#i_5662887

sfan5 commented on builtin/common/serialize.lua:
```diff
@@ -133,11 +133,13 @@ function core.serialize(x)
-			for i, v in ipairs(x) do
-				if not np or not np[i] then
-					vals[#vals + 1] = dump_or_ref_val(v)
+			if vector.is_vector(x) ~= 2 then
```
If this code is required for serialization not to break, that could be another backwards-compatibility problem.

Desour replied:
Modified vector tables that had numerical indexes (starting with 1), additional to x, y and z, won't work anymore anyways. Hence, iteration through the table with `ipairs` makes no sense here.
Things like `local v = vector.new() v[1] = "bla"` can't be preserved, `v[1]` now accesses `x`, which is intended.

sfan5 replied:
I know that ipairs on a vector makes no sense.
The problem here however is if someone has copy-paste serialization code, that code will now serialize vectors incorrectly.

Desour replied:
Such mods (if existent) will have to be fixed then.

Desour commented:
next/pairs now also iterates through new indexes (other than x,y,z).

rubenwardy commented:
Please note that the previous PR to do something similar was rejected due to performance: #6883

Desour commented:
I've fixed the problem with json, see tests in minimal.

Desour commented:
PR is ready btw. (apart from excessive testing).
And thanks for all the reviewing you've done, sfan5!

paramat commented:
Issue #9146 was disapproved.
Issue #7196 has 1 core dev supporting in 2 years, with other core devs not seeing a benefit.
Marked as low priority because the issue is.

From https://github.com/minetest/minetest/issues/7196#issuecomment-593084305
> My pr to do this showed a 4% slowdown. This was with creating a new object each time. I propose an alternative implementation that is class based and does not copy each time

Is this PR class based?
Will there be a reduction of performance?
If there is any reduction of performance at all this is not worth doing as the benefits are minor and questionable (see discussion in issues).
Metatable vectors were previously removed from MT partially for performance reasons.

rubenwardy commented:
>  Issue #7196 has 1 core dev supporting in 2 years, with other core devs not seeing a benefit.

Other core devs being you

paramat commented:
I was including you in that, due to your comment https://github.com/minetest/minetest/issues/7196#issuecomment-378675970 : "It's really not clearer".

rubenwardy commented:
![image](https://user-images.githubusercontent.com/2122943/85903821-2d227d80-b7ff-11ea-8ace-6ff82cb050ec.png)

paramat commented:
Ugh, i misunderstood. Will edit my post.

Desour commented:
Squashed and rebased.


> From [#7196 (comment)](https://github.com/minetest/minetest/issues/7196#issuecomment-593084305)
> 
> > My pr to do this showed a 4% slowdown. This was with creating a new object each time. I propose an alternative implementation that is class based and does not copy each time
> 
> Is this PR class based?

Please explain what you are referring to with "class based". Does lua provide something else than metatables to make vectors more class based?

Does "not copy each time" mean something like add-asign?

If someone has advice (eg. for performance), I'd of course like to hear them. : )


Of course I still need to do benchmarking.

I'm also planning to do some changes, like replacing `vector.new(...)` with something like `vector.zero()`, `vector.copy(v)` and `vector.from_compontents(x, y, z)` or adding `vector.pairs` and dis-recommend using `pairs` or `ipairs` on a vector.

rubenwardy commented on builtin/common/vector.lua:
```diff
@@ -1,58 +1,113 @@
+	elseif key == "z" then
+		return rawget(v, 3)
+	end
+	local func = vector[key]
```
return vector[key]

Desour replied:
Done.

rubenwardy commented:
You can also implement __tostring and __concat like so:

```lua
function meta:__tostring()
	return table.concat(self, ", ")
end

function meta.__concat(a, b)
	return tostring(a) .. tostring(b)
end
```

rubenwardy commented:
Metable vectors are massively easier to use. Array-based vectors are also faster in PUC Lua, but I wasn't able to see any improvement in LuaJIT (it's possible my benchmark was badly designed)

Desour commented:
Added the tostring stuff, made `v:func(...)` make use of being able to do `v[num]` and marked the `v+s` and `v*v ` functionality as deprecated (nearly no-one uses that useless stuff).

rubenwardy commented:
You could avoid the backwards compatibility issues and hacks if you created a new table for this, called Vector

Desour commented:
Added short names like `vector.sub` and in-place functions.

Desour commented:
>You could avoid the backwards compatibility issues and hacks if you created a new table for this, called Vector

How will this avoid backwards compatibility issues? I want the new vectors to be usable in old functions that still use the old vectors. A complete replacement or new thing would be very inconvenient for the modder.

Desour commented:
I've split apart most unrelated changes into separate small PRs to make reviewing and merging those easier:
#10323 #10324 #10326 #10327 #10328 #10329

Desour commented:
Ok, I've done some performance test with functions of a mod that use the vector helpers: `vector.rayIter` and `vector.plane` from `vector_extras` (https://github.com/HybridDog/vector_extras):
```lua
local function meassure(name, func, ...)
	print(name..":")
	local time_start = minetest.get_us_time()
	func(...)
	local time_end = minetest.get_us_time()
	print("time taken: " .. ((time_end - time_start) / 1000) .. " ms")
end

print("benchmarking with vector_extras...")

meassure("rayIter", function()
	local i = 0
	for _ in vector.rayIter(vector.new(1, 2, 3), vector.new(3, 4, 7)) do
		i = i + 1
		if i == 100000 then
			break
		end
	end
end)

meassure("plane", vector.plane, {vector.new(500, 0, 0), vector.new(0, 500, 0), vector.new(0, 0, 500)})

print("benchmarking with vector_extras done.")
```

Output:
(with LuaJIT 2.1.0-beta3, haven't tested with PUC lua yet)
(multiple numbers in one slot means multiple measurements)

branch | rayIter | plane
---|---|---
master | 360.608 ms 367.266 ms | 420.679 ms 421.095 ms
vector_metanonum (#10330: no array-style) | 552.34 ms 545.003 ms | 884.443 ms 886.22 ms
vector_metatable (this PR) | 1159.879 ms 1146.789 ms | 111895.29 ms

As you can see, there's a serious performance impact. :(

Desour commented:
Ok, after the new commits (the second one had no influence):
branch | rayIter | plane
---|---|---
vector_metatable (this PR) | 295.707 ms 295.613 ms | 1038.246 ms 1025.771 ms

I'd love to hear from someone with more knowledge than me about how there's such giant difference difference.

rubenwardy commented:
What's the source code of plane and rayIter? Are you using LuaJIT or Lua?

Desour commented:
> What's the source code of plane and rayIter?

Here are links to the lines:
`rayIter`:
https://github.com/HybridDog/vector_extras/blob/8c7aaf6c0bd5f4fa2ae4d41c22454c3f7d4fe42a/init.lua#L101-L131
`plane`:
https://github.com/HybridDog/vector_extras/blob/8c7aaf6c0bd5f4fa2ae4d41c22454c3f7d4fe42a/init.lua#L274-L327
(I had to do some modifications because of warnings and plane crashing, but my PR there is already merged, so pull from master.)

>Are you using LuaJIT or Lua?

>(with LuaJIT 2.1.0-beta3, haven't tested with PUC lua yet)

Desour commented:
For comparison:
branch | rayIter | plane
---|---|---
PR #6883 | 376.762 ms 382.18 ms | 420.272 ms 420.506 ms

rubenwardy commented:
For my non-MT game, I made a similar implementation to this PR:

https://gist.github.com/rubenwardy/ed715680ead9624aee5a1d14311e8bb4

My game also returns proper Vectors from `:get_pos()` etc, lots of things are possible when you don't need to worry about backwards compatiblity

Desour commented:
After the pre-last commit:
branch | rayIter | plane
---|---|---
vector_metatable | 229.366 ms 228.238 ms | 531.62 ms 537.797 ms

rubenwardy commented:
nice improvement

Desour commented:
Ok, I've also tested for PUC Lua now:
I didn't want to wait too long, so I used smaller numbers for non-jit lua (see details for code).

<details>

```lua
local function meassure(name, func, ...)
	print(name..":")
	local time_start = minetest.get_us_time()
	func(...)
	local time_end = minetest.get_us_time()
	print("time taken: " .. ((time_end - time_start) / 1000) .. " ms")
end

local is_luajit = rawget(_G, "jit") ~= nil

print("benchmarking with vector_extras...")
print("Lua: " .. (is_luajit and "LuaJIT" or "PUC Lua"))

local n1 = is_luajit and 100000 or 50000
meassure("rayIter", function()
	local i = 0
	for _ in vector.rayIter(vector.new(1, 2, 3), vector.new(3, 4, 7)) do
		i = i + 1
		if i == n1 then
			break
		end
	end
end)

local n2 = is_luajit and 500 or 100
meassure("plane", vector.plane, {vector.new(n2, 0, 0), vector.new(0, n2, 0), vector.new(0, 0, n2)})

print("benchmarking with vector_extras done.")

```

</details>

branch | rayIter | plane
---|---|---
master | 1237.238 ms 1261.104 ms | 1935.361 ms 1964.632 ms
vector_metatable | 2297.017 ms 2315.513 ms | 3796.007 ms 3833.966 ms

Desour commented:
- There's now a unittest for each function.
- #10330 was updated (changes of this PR were applied).
- Benchmarking results are now in first comment: https://github.com/minetest/minetest/pull/9612#issue-400529715
- Look there for details.
- Old benchmark result comments were marked as outdated to avoid confusion.
- Benchmarking was done with stable luajit this time. Hence, results may differ.
- Benchmarked code was modified to use the new features. (Both versions were tested.)
- Feel free to try to reproduce the results.
- It looks like #10330 is superior over this PR in nearly all aspects.

Wuzzy2 commented:
While the idea is great, and I would love to have something like that in MT, I'm afraid it cannot be simply approved. Why?

Because in many games, vectors are usually created "by hand", rather than with `vector.new`, i.e. by writing `v = {x=1,y=2,z=3}` directly. And those "vectors" won't magically get the metatables. This will almost certainly lead to compability problems. And I don't you can blame the games/mods for that because `vector.new` literally was just the same as `v = {x=1,y=2,z=3}`, and it is also not officially discouraged practise to do so (as far I know).

So unless you have an idea how to deal with that, I unfortunately have to give a :-1: for that one, based on principle. Although I really hate to do this. :-(

appgurueu commented on src/script/common/c_content.cpp:
```diff
@@ -1770,6 +1770,32 @@ void read_json_value(lua_State *L, Json::Value &root, int index, u8 recursion)
 		size_t len;
 		const char *str = lua_tolstring(L, index, &len);
 		root = std::string(str, len);
+	} else if (is_real_lua_vector(L)) {
```
This is inappropriate; vectors deserve no special handling in JSON. The provided implementation will decrease performance.

Desour commented:
> Because in many games, vectors are usually created "by hand", rather than with `vector.new`, i.e. by writing `v = {x=1,y=2,z=3}` directly. And those "vectors" won't magically get the metatables. This will almost certainly lead to compability problems. And I don't you can blame the games/mods for that because `vector.new` literally was just the same as `v = {x=1,y=2,z=3}`, and it is also not officially discouraged practise to do so (as far I know).

This shouldn't be a problem as long as you only use the metatable features (indexing with numbers, calling functions with :, operators) on vectors where you know that they have a metatable (ie. vectors that you have created in your own code). The vector helper functions (`vector.*`) should still work with the `{x=1,y=2,z=3}` tables.

 I will close this for now.
Storing vectors with array-like tables has too many disadvantages regarding code complexity and hackiness to keep backwards compatibility, and there's no real benefit.
I will probably open a #10330-like PR when I'm more active.
